File: test/TestCases/napi-dotnet/Overloads.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Numerics;
using System.Threading.Tasks;

namespace Microsoft.JavaScript.NodeApi.TestCases;

[JSExport]
public class Overloads
{
    public Overloads()
    {
    }

    public Overloads(int intValue)
    {
        IntValue = intValue;
    }

    public Overloads(string stringValue)
    {
        StringValue = stringValue;
    }

    public Overloads(int intValue, string stringValue)
    {
        IntValue = intValue;
        StringValue = stringValue;
    }

    public Overloads(ITestInterface obj)
    {
        StringValue = obj.Value;
    }

    public int? IntValue { get; private set; }

    public string? StringValue { get; private set; }

    public void SetValue(int intValue)
    {
        IntValue = intValue;
    }

    public void SetValue(string stringValue)
    {
        StringValue = stringValue;
    }

    public void SetValue(int intValue, string stringValue)
    {
        IntValue = intValue;
        StringValue = stringValue;
    }

    public void SetValue(ITestInterface obj)
    {
        StringValue = obj.Value;
    }

    // Method with overloaded name in C# is given a non-overloaded export name.
    [JSExport("setDoubleValue")]
    public void SetValue(double doubleValue)
    {
        IntValue = (int)doubleValue;
    }

    public static string NumericMethod(int value) => $"{value}: int";
    public static string NumericMethod(long value) => $"{value}: long";
    public static string NumericMethod(float value) => $"{value}: float";
    public static string NumericMethod(double value) => $"{value}: double";
    public static string NumericMethod2(string value1, int value2)
        => $"{value1}: string, {value2}: int";
    public static string NumericMethod2(string value1, double value2)
        => $"{value1}: string, {value2}: double";

    public static string NumericMethod3(byte value) => $"{value}: byte";
    public static string NumericMethod3(sbyte value) => $"{value}: sbyte";
    public static string NumericMethod3(ushort value) => $"{value}: ushort";
    public static string NumericMethod3(short value) => $"{value}: short";
    public static string NumericMethod3(uint value) => $"{value}: uint";
    public static string NumericMethod3(int value) => $"{value}: int";
    public static string NumericMethod3(ulong value) => $"{value}: ulong";
    public static string NumericMethod3(long value) => $"{value}: long";

    public static string ClassMethod(ClassObject value) => $"{value.Value}: ClassObject";
    public static string ClassMethod(BaseClass value) => $"{value.Value1}: BaseClass";
    public static string ClassMethod(SubClass value) => $"{value.Value2}: SubClass";
    public static string ClassMethod(StructObject value) => $"{value.Value}: StructObject";

    public static string InterfaceMethod(IBaseInterface value)
        => $"{value.Value1}: IBaseInterface";
    public static string InterfaceMethod(ISubInterface value)
        => $"{value.Value2}: ISubInterface";

    public static string CollectionMethod1(int[] value)
        => $"[{string.Join(", ", value)}]: int[]";
    public static string CollectionMethod1(ICollection<int> value)
        => $"[{string.Join(", ", value)}]: ICollection<int>";
    public static string CollectionMethod1(IList<int> value)
        => $"[{string.Join(", ", value)}]: IList<int>";
    public static string CollectionMethod1(ISet<int> value)
        => $"[{string.Join(", ", value)}]: ISet<int>";
    public static string CollectionMethod1(IDictionary<int, int> value)
        => $"[{string.Join(", ", value)}]: IDictionary<int, int>";

    public static string CollectionMethod2(IEnumerable<int> value)
        => $"[{string.Join(", ", value)}]: IEnumerable<int>";
    public static string CollectionMethod2(IReadOnlyCollection<int> value)
        => $"[{string.Join(", ", value)}]: IReadOnlyCollection<int>";
    public static string CollectionMethod2(IReadOnlyList<int> value)
        => $"[{string.Join(", ", value)}]: IReadOnlyList<int>";
    public static string CollectionMethod2(IReadOnlyDictionary<int, int> value)
        => $"[{string.Join(", ", value)}]: IReadOnlyDictionary<int, int>";

    public static string CollectionMethod3(IEnumerable<int> value)
        => $"[{string.Join(", ", value)}]: IEnumerable<int>";
    public static string CollectionMethod3(ICollection<int> value)
        => $"[{string.Join(", ", value)}]: ICollection<int>";

    public static Task<string> CollectionMethod4(IEnumerable<int> value)
        => Task.FromResult($"[{string.Join(", ", value)}]: IEnumerable<int>");
    public static async Task<string> CollectionMethod4(IAsyncEnumerable<int> value)
    {
        List<int> list = new();
        await foreach (var item in value)
        {
            list.Add(item);
        }
        return $"[{string.Join(", ", list)}]: IAsyncEnumerable<int>";
    }

    public static string DateTimeMethod(DateTime value) => $"{value:s}: DateTime";
    public static string DateTimeMethod(DateTimeOffset value) => $"{value:s}: DateTimeOffset";
    public static string DateTimeMethod(TimeSpan value) => $"{value}: TimeSpan";

    public static string OtherMethod(TestEnum value) => $"{value}: TestEnum";
    public static string OtherMethod(Guid value) => $"{value}: Guid";
    public static string OtherMethod(BigInteger value) => $"{value}: BigInteger";
    public static string OtherMethod(Task value) => $"Task";
    public static string OtherMethod(TestDelegate value) => $"{value("test")}: TestDelegate";

    public static string NullableNumericMethod(int? value) =>
        $"{(value == null ? "null" : value.ToString())}: int?";
    public static string NullableNumericMethod(double value) => $"{value}: double";
}

File: test/TestCases/napi-dotnet/overloads.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

const assert = require('assert');

/** @type {import('./napi-dotnet')} */
const binding = require('../common').binding;

const Overloads = binding.Overloads;
const ClassObject = binding.ClassObject;
const BaseClass = binding.BaseClass;
const SubClass = binding.SubClass;
const StructObject = binding.StructObject;
const TestEnum = binding.TestEnum;

// Overloaded constructor
const emptyObj = new Overloads();
assert.strictEqual(emptyObj.intValue, undefined);
assert.strictEqual(emptyObj.stringValue, undefined);

const intObj = new Overloads(1);
assert.strictEqual(intObj.intValue, 1);
assert.strictEqual(intObj.stringValue, undefined);

const stringObj = new Overloads('two');
assert.strictEqual(stringObj.intValue, undefined);
assert.strictEqual(stringObj.stringValue, 'two');

const comboObj = new Overloads(3, 'three');
assert.strictEqual(comboObj.intValue, 3);
assert.strictEqual(comboObj.stringValue, 'three');

const objValue = new ClassObject();
objValue.value = 'test';
const objFromClass = new Overloads(objValue);
assert.strictEqual(objFromClass.stringValue, 'test');

// Overloaded method with basic resolution by parameter count and JS type.
const obj1 = new Overloads();
obj1.setValue(1);
assert.strictEqual(obj1.intValue, 1);
assert.strictEqual(obj1.stringValue, undefined);

const obj2 = new Overloads();
obj2.setValue('two');
assert.strictEqual(obj2.intValue, undefined);
assert.strictEqual(obj2.stringValue, 'two');

const obj3 = new Overloads();
obj3.setValue(3, 'three');
assert.strictEqual(obj3.intValue, 3);
assert.strictEqual(obj3.stringValue, 'three');

const obj4 = new Overloads();
obj4.setValue(objValue);
assert.strictEqual(obj4.stringValue, 'test');

// Overloaded C# method with explicit JS method name.
const obj5 = new Overloads();
obj5.setDoubleValue(5.0);
assert.strictEqual(obj5.intValue, 5);

// Overloaded method with resolution by matching numeric type.
assert.strictEqual(Overloads.numericMethod(1), '1: int');
assert.strictEqual(Overloads.numericMethod(10000000000), '10000000000: long');
assert.strictEqual(Overloads.numericMethod(1.11), '1.11: double');
assert.strictEqual(Overloads.numericMethod2('test', 2), 'test: string, 2: int');
assert.strictEqual(Overloads.numericMethod2('test', 2.22), 'test: string, 2.22: double');

// Overloaded method with resolution by selecting best numeric type specificity.
assert.strictEqual(Overloads.numericMethod3(1), '1: byte');
assert.strictEqual(Overloads.numericMethod3(-1), '-1: sbyte');
assert.strictEqual(Overloads.numericMethod3(1000), '1000: ushort');
assert.strictEqual(Overloads.numericMethod3(-1000), '-1000: short');
assert.strictEqual(Overloads.numericMethod3(1000000), '1000000: uint');
assert.strictEqual(Overloads.numericMethod3(-1000000), '-1000000: int');
assert.strictEqual(Overloads.numericMethod3(10000000000), '10000000000: ulong');
assert.strictEqual(Overloads.numericMethod3(-10000000000), '-10000000000: long');

// Overloaded method with resolution by matching object type.
assert.strictEqual(Overloads.classMethod(new ClassObject('class')), 'class: ClassObject');
assert.strictEqual(Overloads.classMethod(new BaseClass(1)), '1: BaseClass');
assert.strictEqual(Overloads.classMethod(new SubClass(1, 2)), '2: SubClass');
assert.strictEqual(Overloads.classMethod({ value: 'struct' }), 'struct: StructObject');
assert.strictEqual(Overloads.classMethod(new StructObject('struct2')), 'struct2: StructObject');

// Overloaded method with resolution by matching interface type.
assert.strictEqual(Overloads.interfaceMethod(new BaseClass(1)), '1: IBaseInterface');
assert.strictEqual(Overloads.interfaceMethod(new SubClass(1, 2)), '2: ISubInterface');

// Overloaded method with resolution by matching collection type.
assert.strictEqual(Overloads.collectionMethod1([1, 2, 3]), '[1, 2, 3]: IList<int>');
assert.strictEqual(Overloads.collectionMethod1(new Set([1, 2, 3])), '[1, 2, 3]: ISet<int>');
assert.strictEqual(Overloads.collectionMethod1(
    new Map([[1, 10], [2, 20], [3, 30]])), '[[1, 10], [2, 20], [3, 30]]: IDictionary<int, int>');

// Overloaded method with resolution by matching read-only collection type.
assert.strictEqual(Overloads.collectionMethod2(
    [1, 2, 3]), '[1, 2, 3]: IReadOnlyList<int>');
assert.strictEqual(Overloads.collectionMethod2(
    new Set([1, 2, 3])), '[1, 2, 3]: IReadOnlyCollection<int>');
assert.strictEqual(Overloads.collectionMethod2(
    new Map([[1, 10], [2, 20], [3, 30]])),
    '[[1, 10], [2, 20], [3, 30]]: IReadOnlyDictionary<int, int>');

// Overloaded method with resolution by matching iterable or collection type.
const testIterable = {
    [Symbol.iterator]: function* () {
        yield 1;
        yield 2;
        yield 3;
    }
};
const testAsyncIterable = {
    [Symbol.asyncIterator]: async function* () {
        yield 1;
        yield 2;
        yield 3;
    }
};
assert.strictEqual(Overloads.collectionMethod3(
    testIterable), '[1, 2, 3]: IEnumerable<int>');
assert.strictEqual(Overloads.collectionMethod3(
    [1, 2, 3]), '[1, 2, 3]: ICollection<int>');
assert.strictEqual(Overloads.collectionMethod3(
    new Set([1, 2, 3])), '[1, 2, 3]: ICollection<int>');
Overloads.collectionMethod4([1, 2, 3]).then(
    (result) => assert.strictEqual(result, '[1, 2, 3]: IEnumerable<int>'));
Overloads.collectionMethod4(testIterable).then(
    (result) => assert.strictEqual(result, '[1, 2, 3]: IEnumerable<int>'));
Overloads.collectionMethod4(testAsyncIterable).then(
    (result) => assert.strictEqual(result, '[1, 2, 3]: IAsyncEnumerable<int>'));

// The following types have special marshalling behaviors:
// DateTime(Offset), TimeSpan, Enums, Guid, BigInteger, Task, Delegate

const dateWithKind = new Date(Date.UTC(2024, 1, 29));
dateWithKind.kind = 'utc';
assert.strictEqual(Overloads.dateTimeMethod(dateWithKind), '2024-02-29T00:00:00: DateTime');
const dateWithOffset = new Date(2024, 1, 29);
dateWithOffset.offset = -10 * 60;
assert.strictEqual(Overloads.dateTimeMethod(dateWithOffset), '2024-02-29T00:00:00: DateTimeOffset');
assert.strictEqual(Overloads.dateTimeMethod(11 * 60 * 1000), '00:11:00: TimeSpan');

assert.strictEqual(Overloads.otherMethod(TestEnum.One), 'One: TestEnum');
assert.strictEqual(Overloads.otherMethod('00000000-0000-0000-0000-000000000000'),
    '00000000-0000-0000-0000-000000000000: Guid');
assert.strictEqual(Overloads.otherMethod(1000000000000000000000000n),
    '1000000000000000000000000: BigInteger');
assert.strictEqual(Overloads.otherMethod(BigInt('1000000000000000000000000')),
    '1000000000000000000000000: BigInteger');
assert.strictEqual(Overloads.otherMethod(Promise.resolve()), 'Task');
assert.strictEqual(Overloads.otherMethod((value) => value.toUpperCase()), 'TEST: TestDelegate');

// Overloaded method with resolution of null / undefined parameters.
assert.strictEqual(Overloads.nullableNumericMethod(null), 'null: int?');
assert.strictEqual(Overloads.nullableNumericMethod(undefined), 'null: int?');
assert.strictEqual(Overloads.nullableNumericMethod(3), '3: int?');
assert.strictEqual(Overloads.nullableNumericMethod(4.4), '4.4: double');
